COMPAT: Fix indent level bug preventing wrapper function rename

matthagy opened the pull request:
Original code intends to rename the wrapper function f using the provided name, but this isn't happening because code is incorrectly indented an extra level.
```python
Example:
>>> from pandas.core.groupby import GroupBy
>>> GroupBy.sum.__name__
'f'
```

Should be 'sum'.

Commit:
Fix indent level bug preventing wrapper function rename

Original code intends to rename the wrapper function f using the provided name, but this isn't happening because code is incorrectly indented an extra level.

Example:
>>> from pandas.core.groupby import GroupBy
>>> GroupBy.sum.__name__
'f'

Should be 'sum'.

jreback commented:
can u add some tests for this (and validate all functions names in the whitelist would be great)

Commit:
Test renaming of _groupby_function wrapper function

matthagy commented:
@jreback, tried adding a test for all whitelist function names and discovered there are additional places where attribute and method names are inconsistent. Replaced that with a test that only covers the names corresponding to this current bug fix. If you think it makes sense to hold off until all whitelist function names are properly handled then I can close this PR for now and open an issue.

jreback commented:
just make it 2. tests
1 for good ones
1 for failing ones
then can add a TODO for the failing ones

Commit:
Test for consistency of attribute and method names

Commented out and marked with a TODO since some are currently inconsistent
and not immediately obvious how to fix all of them.

matthagy commented:
@jreback, added back failing test, commented out, and marked w/ TODO. Does reuse a an existing test to prevent duplicating whitelists or refactoring them to be class members. Let me know if you had something else in mind.

jreback commented on pandas/tests/test_groupby.py:
```diff
@@ -5990,7 +5990,9 @@ def test_groupby_whitelist(self):
-                getattr(type(gb), m)
+                f = getattr(type(gb), m)
+                # TODO: Fix inconsistencies between attribute and method names
+                # self.assertEqual(f.__name__, m)
```
can u made an actual test of things in the whitelist which fail

Commit:
Revise attribute/method consistency check to skip known inconsistencies

matthagy commented:
@jreback, revised test to show things in whitelist that currently fail. Test now also handles all of the other whitelist cases that aren't known issues.

jreback commented on pandas/tests/test_groupby.py:
```diff
@@ -5986,11 +5986,44 @@ def test_groupby_whitelist(self):
             # 'nlargest', 'nsmallest',
         ])
 
+        # TODO: Fix these inconsistencies between attribute and method names
```
what does this comment mean?

jreback commented on pandas/tests/test_groupby.py:
```diff
@@ -5986,11 +5986,44 @@ def test_groupby_whitelist(self):
+            'unique'
+        ])
+
         for obj, whitelist in zip((df, s), (df_whitelist, s_whitelist)):
```
what I mean is create *another* function where you assert that these are NOT equal (with the sub-list of those that are broken), so that when they are fixed this tests will break (and the list will need to be updated). IOW, create a list like the white-list of the broken ones, have the existing test just skip on them (as you do below), but then create *another* function that asserts that the broken ones are not equal.

Commit:
Added a test for known inconsistent attribute/method names

matthagy commented:
@jreback, created the additional test to assert that method `__name__` is inconsistent with attribute name for these known issues. Also revised the comment explaining the inconsistencies.

jreback commented:
@matthagy can you rebase / update.

sorry got a bit lost in the shuffle.

matthagy commented:
@jreback, rebased.

jreback commented:
can you rebase?

matthagy commented:
@jreback, rebased.

jreback commented:
thanks @matthagy 
lgtm. ping on green.

jreback commented:
can you rebase / update

jreback commented:
@matthagy can you rebase / ping on green.

jreback committed:
Move _groupby_function inside GroupBy

Add support for __qualname__

jreback commented:
@matthagy I pushed a change here to make this also set ``__qualname__`` (we do this already all Series/DataFrame anyhow, so seemed easy to add here.

Can you push a whatsnew note, otherwise this lgtm. I think you had several TODO's in the tests for functions which don't set things properly, can you have a look at those?

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -3753,6 +3753,19 @@ def test_groupby_selection_with_methods(self):
                            g_exp.filter(lambda x: len(x) == 3))
 
+    # The methods returned by these attributes don't have a __name__ attribute
+    # that matches that attribute.
```
these here

jreback commented on pandas/tests/groupby/test_groupby.py:
```diff
@@ -3929,6 +3986,12 @@ def test_tab_completion(self):
              'ffill', 'bfill', 'pad', 'backfill', 'rolling', 'expanding'])
         self.assertEqual(results, expected)
 
+    def test_groupby_function_rename(self):
```
is this test superfluous? (or just a quick test)?

jreback commented:
cc @mrocklin

jreback committed:
PEP

jreback committed:
doc

jreback committed:
clean/reorg tests
